src/gha_runner/clouddeployment.py

```diff
@@ -0,0 +1,235 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+import importlib.resources
+import boto3
+from string import Template
+
+
+class CloudDeployment(ABC):
+    """Abstract base class for cloud deployment.
+
+    This class defines the interface for cloud deployment classes.
+
+    """
+
+    @abstractmethod
+    def create_instance(self, count: int) -> list[str]:
+        """Create instances in the cloud provider and return their IDs.
+
+        Parameters
+        ----------
+        count : int
+            The number of instances to create.
+
+        Returns
+        -------
+        list[str]
+            A list of instance IDs.
+
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def remove_instances(self, ids: list[str]):
+        """Remove instances from the cloud provider.
+
+        Parameters
+        ----------
+        ids : list[str]
+            A list of instance IDs to remove.
+
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def wait_until_ready(self, ids: list[str], **kwargs):
+        """Wait until instances are in a ready state.
+
+        Parameters
+        ----------
+        ids : list[str]
+            A list of instance IDs to wait for.
+        **kwargs : dict, optional
+            Additional arguments to pass to the waiter.
+
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def wait_until_removed(self, ids: list[str], **kwargs):
+        """Wait until instances are removed.
+
+        Parameters
+        ----------
+        ids : list[str]
+            A list of instance IDs to wait for.
+        **kwargs : dict, optional
+            Additional arguments to pass to the waiter.
+
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def instance_running(self, id: str) -> bool:
+        """Check if an instance exists.
+
+        Parameters
+        ----------
+        id : str
+            The instance ID to check.
+
+        Returns
+        -------
+        bool
+            True if the instance exists, False otherwise.
+
+        """
+        raise NotImplementedError
+
+
+@dataclass
+class AWS(CloudDeployment):
+    image_id: str
+    instance_type: str
+    gh_runner_token: str
+    home_dir: str
+    repo: str
+    tags: list[dict[str, str]]
+    region_name: str
+    runner_release: str
+    labels: str = ""
+    subnet_id: str = ""
+    security_group_id: str = ""
+    iam_role: str = ""
+    script: str = ""
+
+    def __build_aws_params(self, count: int, user_data_params: dict) -> dict:
```

Comment:
This is a bit of overkill: when I suggested making it private, I meant in the sense of a single leading `_`, which gives the guideline "this isn't intended for public consumption."

Making it truly class-specific via name-mangling probably isn't needed or desired (its main use, as far as I know, is as a trick to sort of break inheritance, which we're not trying to do here -- TBH, I don't think I've ever had a real-world situation where that was the solution I needed.)

Reply:
Fixed.